define widget type mimicking actions with flexible lua scripts

dterrahe opened the pull request:
This PR has picked up a bunch of separate functionalities with interdependencies (so splitting into separate PRs would involve extra work, waiting and rebasing). It might be best just to list some release notes entries:

- the lua call to `dt.gui.action` has become more flexible, with most parameters optional, so you can read the focused status of a module doing just `dt.gui.action("iop/filmicrgb", "focus")`
- tooltips show these compacter lua commands in mapping mode (only adding the last parameter, _instance_, if the module supports multi-instance) and have been added to presets and styles menus as well
- lua commands can be copied to the clipboard, using ctrl+v in the shortcuts dialog, both from a selected action or shortcut, or long right click when in mapping mode (over a widget) or in the presets/styles menus
- the shown/copied lua command for a slider or combobox will set the value it currently has
- a bug with shortcuts (and `dt.gui.action`) to set the active item in a combobox with varying content has been fixed and now it is also possible to directly set the values of the combos for the focused module's blending mode etc. (by setting the shortcuts _effect_)
- added section headers to the _sort by_ drop-down (_files_, _times_, etc)
- shortcuts assigned to presets or styles will be shown when hovering over them in their menu
- _long_ left clicking a preset will keep the menu open so you can quickly switch between several to see the effect without having to repeatedly click the preset button to reopen the menu. You can also scroll over the preset button to switch to previous/next presets (like you already could via shortcuts)
- a shortcut can now be directed to a lua script that mimics a standard slider, dropdown or button, but dynamically selects the real widget(s) that receive(s) it, based on for example which module is focused or enabled. The advantage is that all fallbacks work as normal, so you can assign a midi knob to it and turning it (holding shift/ctrl to speed up/down) or pressing it to reset work regardless of which widget receives it. Basically this is a much more flexible alternative to the fake widgets under `processing modules/<focused>`. This allows owners of for example an x-touch mini to use their scarce rotors in different, fully configurable, ways while working in different modules (which can also be focused using midi buttons which will then light up). Such configurations could be shared using https://github.com/darktable-org/lua-scripts
- support shortcuts to sliders/combos created in lua, either via visual mapping mode or in the shortcuts dialog under the _lua_ category. _Elements_ and _effects_ are not supported.

ORIGINAL POST:
Introduce a new lua call `dt.gui.mimic` that creates an action for a "fake" widget (in the shortcuts action list under "lua scripts"). The widget _looks_ like a normal _slider_ or _dropdown_, so when a shortcut is mapped to it, it accepts the same elements and effects and uses the same fallbacks. But instead of sending these directly to a _real_ widget, it calls a lua routine, which can then decide (based on for example which module has the focus, but this could be anything) to which _real_ widget the shortcut should be forwarded. It does this by calling `dt.gui.action` with the arguments it received.

@jenshannoschwalm You could do what you mentioned in https://github.com/darktable-org/darktable/pull/11361#issuecomment-1152726578 but instead of having different modifiers direct where the _contrast_ knob goes, it could depend on whether filmic is currently focused. And shift and ctrl would still work to speed it up or slow it down.

```
local dt = require "darktable"

dt.gui.mimic("slider", "multi-contrast",
  function(action, element, effect, size)
      if dt.gui.action("iop/filmicrgb", "focus") == 1 then
        which = "iop/filmicrgb/contrast"
      else
        which = "iop/colorbalancergb/contrast"
      end
      return dt.gui.action(which, element, effect, size)
    end
)
```

You could also have one shortcut be sent to two widgets at the same time
```
dt.gui.mimic("dropdown", "double combo",
  function(action, element, effect, size)
    dt.gui.action("iop/filmicrgb/contrast in shadows", element, effect, size)
    return dt.gui.action("iop/filmicrgb/contrast in highlights", element, effect, size)
  end
)
```
In this case, if mapped to a knob on the x-touch mini, the rotator lights would indicate the status of the second widget.

This could be expanded, for the x-touch or similar, to define 8 fake sliders and assign all of the rotator knobs to those. They could then be redirected to the different point on the curve in color zones, if that module is active, or to the different parameters of shapes, if a mask is currently being edited (test `dt.gui.action("iop/blend/tools/show and edit mask elements", 0, "", "")`), and to some other frequently used sliders when neither of these are the case.
Old style calls are still supported so existing scripts shouldn't break.

EDIT: makes calls to dt.gui.action more flexible and moves _instance_ parameter to the end (since it is only relevant for a subset of actions (multi-instance iop modules) and even then usually defaults to 0. Most parameters are now optional, so you can read the focus state of a module with `dt.action.gui("iop/modulename", "focus")` or set it with `dt.action.gui("iop/modulename", "focus", 1)`

EDIT2: fixes a bug with combobox shortcuts/dt.gui.action calls that specify a specific "item:" in cases where the list in the live combo options was manipulated at runtime. The nth item in the combo would not correspond to the nth item in the actions list anymore. Now more robust. Could be used to provide a static list of options in, say, the blend mode dropdown as well to allow them to be selected by shortcut.

dterrahe commented:
x-touch mini users may want to try out the below. If color zones is focused, the knobs will work on that. Otherwise if the _currently focused_ module is in mask editing mode, the shape parameters will be adjusted. The idea can be expanded and customized.
```
local dt = require "darktable"

function xtouchknob(action, element, effect, size)
  k = tonumber(action:sub(-1))

  if dt.gui.action("iop/blend/tools/show and edit mask elements", 0, "") ~= 0 then
    local s = { "opacity", "size", "feather", "hardness","rotation","curvature","compression" }
    which = "lib/masks/properties/" .. s[k]

  elseif dt.gui.action("iop/colorzones", 0, "focus") ~= 0 then
    local e = { "red", "orange", "yellow", "green", "aqua", "blue", "purple", "magenta" }
    element = e[k]
    which = "iop/colorzones/graph"

  elseif k == 4 and dt.gui.action("iop/colorbalancergb", 0, "focus") ~= 0 then
    which = "iop/colorbalancergb/contrast"

  else
    local s = { "iop/exposure/exposure",
                "iop/filmicrgb/white relative exposure",
                "iop/filmicrgb/black relative exposure",
                "iop/filmicrgb/contrast",
                "iop/crop/left",
                "iop/crop/right",
                "iop/crop/top",
                "iop/crop/bottom" }
    which = s[k]
  end
  return dt.gui.action(which, 0, element, effect, size)
end

for k = 1,8 do
  dt.gui.mimic("slider", "x-touch knob ".. k, xtouchknob)
end
```

Import the following into your shortcuts (after backing up your current ones):
```
None;midi:CC1=lua/x-touch knob 1
None;midi:CC2=lua/x-touch knob 2
None;midi:CC3=lua/x-touch knob 3
None;midi:CC4=lua/x-touch knob 4
None;midi:CC5=lua/x-touch knob 5
None;midi:CC6=lua/x-touch knob 6
None;midi:CC7=lua/x-touch knob 7
None;midi:CC8=lua/x-touch knob 8
midi:C#0=iop/colorzones;focus
midi:D0=iop/colorbalancergb;focus
midi:D#0=iop/blend/tools/show and edit mask elements
```
This also maps a few buttons to toggle (and indicate!) the focus of the relevant modules.

dterrahe commented:
@wpferguson had I known how easy it is to have default parameters in the call to dt.gui.action, I would have placed the _instance_ at the end, because it is rarely used.
`dt.gui.action("iop/colorzones", 0, "focus")` in the above could then be simplified to
`dt.gui.action("iop/colorzones", "focus")`
or to toggle it
`dt.gui.action("iop/colorzones", "focus", "", 1)`

I presume it is too late now to make such an API change? wdyt

wpferguson commented:
> I presume it is too late now to make such an API change? wdyt

I think there aren't many people using `dt.gui.action()` in scripts right now, so it should be too big an impact.  We could get the word out on pixls.us, the mailing lists, and the docs.

wpferguson commented:
I'll test this and see what it can do :smiley:

dterrahe commented:
For tone equalizer add this before the last "else" in the script:
```
  elseif dt.gui.action("iop/toneequal", 0, "focus") ~= 0 then
    which ="iop/toneequal/simple/"..(k-9).." EV"
```
(this doesn't use the "+0" slider; x-touch only has 8 knobs.

dterrahe commented:
> @wpferguson had I known how easy it is to have default parameters in the call to dt.gui.action, I would have placed the _instance_ at the end, because it is rarely used.

I have now made this change and made all except the first argument (the action itself) optional while still supporting the previous order (as long as in that case _all_ parameters are specified, which used to be a requirement).

wpferguson commented:
Started some testing.  I used the first example above and assigned it a shortcut of a+scroll.  When I tried it in darkroom on an image, a+scroll just zoomed the image in and out.

I assigned c+scroll to the shortcut and it works, adjusting the filmic contrast.  Not sure why a+scroll didn't work.

I was running the -d lua flag and got warnings about locking in the GUI thread.

```
   273.8901 LUA WARNING locking from the gui thread should be avoided
   274.2736 LUA WARNING locking from the gui thread should be avoided
   274.3490 LUA WARNING locking from the gui thread should be avoided
```

Not really sure this is an issue since we are controlling the GUI.  I think the warning is geared toward a lua job that should be running the the lua thread hanging the GUI.

dterrahe commented:
> When I tried it in darkroom on an image, a+scroll just zoomed the image in and out.

By default, `a` is mapped to the "force pan/zoom/rotate with mouse" action, which means scrolling will zoom even if it otherwise temporarily would do something else, like for example resize a mask shape. That overrides any later shortcuts starting with `a` you may define (except if you were to include a modifier like _ctrl_ or _shift_)

> got warnings about locking in the GUI thread

Yes, I almost forgot about that. The call to `dt.gui.mimic` callbacks needs to provide a return value and _should_ just be used to loop back into `dt.gui.action` calls, so there's no point handing over to the separate lua thread (and back into the gui one from there) and then blocking until receiving a signal with the return signal or something massively complicated like that. So I would suggest removing that warning. But of course it _does_ mean if someone writes a misbehaving script, they could easily hang the whole program. That's the nature of things. They should just not do that; this callback is not for doing any actual _processing_.

wpferguson commented:
> By default, a is mapped to the "force pan/zoom/rotate with mouse" action

It works well :smile:

> So I would suggest removing that warning

How about if we move the `#ifdef _DEBUG` that's below it to just above the `if` statement.  That way if we need to debug it's still there, but we've moved it into the more serious level of debugging and out of the normal lua "tell me what you're doiing".

dterrahe commented:
Sounds good to me! I'll make that change.

dterrahe commented:
@TurboGit this is now done (bar bug-fixes if reported). As you can see from the release notes items in the updated first post, this PR has expanded a little. There's some bug fixes and additional functionality, especially on the lua/shortcuts/tooltips/clipboard side.

But there's one new feature, _long_ clicking a preset to keep the menu open, that might (or _not_!) appeal to everybody. Feedback as always welcome.

dterrahe commented:
rebased.

@TurboGit forgot to mention another small cleanup where the code contained some "special" (no-break) spaces that are indistinguishable from normal spaces, except of course when you use them in lua code. https://discuss.pixls.us/t/working-with-modules-from-lua-scripts-in-darktable/32934/83

dterrahe commented:
@wpferguson now that I've got your attention...

A while back we discussed how to make it possible to set up shortcuts to combos and sliders created by lua. I just started thinking about that a little again. The problem is that normally, the lib or iop that the bauhaus widget is going to be put into is known at the moment the widget is created, so it is passed to the `_new` call and stored inside the bauhaus, so it can be used when a label is assigned to at the same time set up the complete action string. In _lua_ the widgets are created first, then the containers and finally they are attached to the lib (which can be created from lua or maybe already existed somewhere). So now we would have to work back somehow at the moment the lib is finalised to find which widgets were put into it (and if any added later, too bad) and then fix them up. And for existing libs, they were already finalised _before_ lua even started. So alternatively maybe we could listen to the "map" signal for the widget, which means as soon as it appears on screen, we could try to work our way back. All very tedious and finicky and if you want to keep the module hidden but still use the shortcut, it is probably not going to work anyway.

So I'm wondering if there is value in a much simpler approach; just adding the actions all directly under the "lua" top level action. That can be achieved by just adding one line for each type. Fake diff below.
```
+++ b/src/lua/widget/combobox.c
@@ -83,6 +83,7 @@ static int label_member(lua_State *L)
   if(lua_gettop(L) > 2) {
     char tmp[256];
     luaA_to(L,char_256,&tmp,3);
+    DT_BAUHAUS_WIDGET(combobox->widget)->module = &darktable.control->actions_lua;
     dt_bauhaus_widget_set_label(combobox->widget,NULL,tmp);
     return 0;
   }
+++ b/src/lua/widget/slider.c
@@ -48,6 +48,7 @@ static int label_member(lua_State *L)
   if(lua_gettop(L) > 2) {
     char tmp[256];
     luaA_to(L,char_256,&tmp,3);
+    DT_BAUHAUS_WIDGET(slider->widget)->module = &darktable.control->actions_lua;
     dt_bauhaus_widget_set_label(slider->widget,NULL,tmp);
     return 0;
   }
```
The obvious drawback of course would be that if you give two widgets in different modules the same name, only one of them will receive the shortcut.

If we somehow can receive the name of the currently running script form the lua_State, we could use that as a _section_ to at least not overlap with other scripts or lua commands.

All the while you have to remember that because the shortcuts are loaded before lua runs, we can't know at that time what _type_ a shortcut refers to, so any _elements_ or _effects_ that were set up and saved would be silently dropped at each startup, so we won't allow them. Only the default for each shortcut will work (and fallbacks).

What do you think? Shall I add those two lines to this PR (maybe including the script name if available)?

wpferguson commented:
> @wpferguson now that I've got your attention...

:rofl:

I've been trying for 2 weeks to find a couple of hours to focus and work on an old ruby program I wrote several years ago.  Today I finally managed some quite time to do it.  As soon as I was finished, I read this...  :smiley: 

The usual widget creation process is 
- create the widgets (buttons, combobox, entry)
- pack and arrange them
- pass them as an argument to the lib or storage creation process

The 2 exceptions are adding a button to the `selected images` or `select` modules.

Widgets in Lua can be much more dynamic.  For instance when you are flipping through pages of scripts in script_manager it's actually just one set of buttons that I remove the labels from, then write new labels and rewrite the callback function.  Comboboxes can be rewritten on the fly also.  But there are other scripts where the widgets are static, such as enfuseAdvanced.

Let me see if I can figure out a way to get the name of the script that's creating the widget.

dterrahe commented:
If I simply add this line
`running_script = script.name`
to `script_manager.lua`, just before the call to `prequire` and then change the calls in `label_member` described above to
```
    lua_getglobal(L, "running_script");
    DT_BAUHAUS_WIDGET(slider->widget)->module = &darktable.control->actions_lua;
    dt_bauhaus_widget_set_label(slider->widget,lua_tolstring(L,-1,NULL),tmp);
```
(and same for combo) then they end up in their own nice sections below `lua` in the shortcuts dialog.

Any script directly run from luarc or not "officially" started from script manager still gets their widgets dumped in the main `lua` section (which I think is "fine").

Separate question (not for this PR): if I look through the lua docs, then if understand correctly, the use of the "require" call means that lua will take care of not running a script twice. Is that maybe the only reason that if I disable and then reenable a script in the manager it still does not load an updated version until restart? Might it be possible (and possibly _better_) to use `loadfile` instead?

wpferguson commented:
> If I simply add this line
running_script = script.name

That's easy enough to do 

> Any script directly run from luarc or not "officially" started from script manager still gets their widgets dumped in the main lua section (which I think is "fine").

Agreed

> Separate question 

One of the goals of script_manager was to be able to stop a script.  I've added a lot of `destroy_` functions to the API in order to achieve this.  The only thing I haven't managed to stop yet is a `lib`, so I just hied it when you turn it off.  If I stop a script, it is still known to lua in the `package.loaded` table.  Simply doing a `package.locaded[script.name] = nil` unloads it and thus forces it to be reloaded on the next start.  This might be a good change to add so that developers wont have to restart darktable when they have a misbehaving script in order to get it to reload.

dterrahe commented:
> > running_script = script.name
> 
> That's easy enough to do

Are you OK with this suggested variable name or would you like to propose something else? I'll then include the c-side in this PR and you can add it to the script at your convenience. Neither side suffers serious damage if the introduction and use of the variable isn't synced.

> This might be a good change to add so that developers wont have to restart darktable when they have a misbehaving script in order to get it to reload.

Yes, that would be great!

wpferguson commented:
> Are you OK with this suggested variable name or would you like to propose something else?

I would make it `sm_running_script`.  `running_script` is kind of generic and the `sm_` will give us some idea of where it's being set.  Or, for that matter you could make it `script_manager_running_script` which would remove all doubt.

dterrahe commented:
`script_manager_running_script` it is

dterrahe commented:
@AlicVB @TurboGit now that shortcuts with randomly ordered combos are more robust, it is possible to add sections to the sort order combo (and generally make some simplification). I've reordered the items to make more logical categories, but am a little struggling with _aspect ratio_ (and obviously don't want to put it in a category of its own). Any suggestions to improve on this welcome!
![image](https://user-images.githubusercontent.com/1549490/220201384-eb389188-3a6d-475f-aa35-bcd3247773a0.png)

dterrahe commented:
> `script_manager_running_script` it is

done

@TurboGit now _really_ finished with this PR ;-)

wpferguson commented:
just tested shortcuts for lua widgets and it works fine :)

dterrahe commented:
Fixed and rebased; now also compiling _without_ USE_LUA. Apologies; I assumed there were problems with CI given the number of PRs.

I'm now getting `corrupted size vs. prev_size` from Linux.latest.LLVM14.build.Release.Ninja trying to run after succesful compile. I'm not seeing this error locally.

dterrahe commented:
Rebased against latest formatting changes.

TurboGit commented:
> @TurboGit Not sure why retriggering that particular CI job fixed it;

Don't know either but I see sometime network issue where the package manager cannot update the machine or plain crash during machine setup. Nothing we can do except monitor and restart the jobs... and crossing fingers :)

TurboGit commented:
> Any suggestions to improve on this welcome!

I saw this question yesterday... and have no better proposal. Let's keep this in files section for now.

TurboGit reviewed (changes requested):
Tested / checked and works on my side. There is multiple features there and I have not checked the Lua ones already approved by @wpferguson. I have some comments and changes proposed.

wpferguson commented on src/lua/lua.c:
```diff
@@ -132,13 +132,13 @@ void dt_lua_init_lock()
 
 void dt_lua_lock_internal(const char *function, const char *file, int line, gboolean silent)
 {
+#ifdef _DEBUG
```
Yes

dterrahe commented:
I reverted some of the changes to the bauhaus API and instead expanded the new `dt_bauhaus_combobox_add_introspection` to take a _start_ and _end_ item to add to a combo, so it is easier to insert section heading in between.

I've used that, in a separate commit, to simplify the filling of the blending combos and to align them more with eachother (all non-raws now have a "chromaticity & lightness" section, for example) and with the list in the shortcut _effects_ dropdown so it is hopefully easier to navigate. This did mean reordering some items, so if that is controversial this commit could be dropped. I've also removed the " modes" ending of each section header since it is repetitive and makes some headings very long which imho doesn't help readability.
![image](https://user-images.githubusercontent.com/1549490/220499573-c38d34a4-b93c-4add-a719-2aebbdd5eb82.png)![image](https://user-images.githubusercontent.com/1549490/220497178-c97d7fd0-d219-47ce-a0d9-642c2a672708.png)

While doing this I noticed that "Lab L-channel" isn't offered anywhere, so I moved it to the "deprecated" section of the list (i.e. it appears at the bottom of the _effects_ list). If this is incorrect, please let me know and I'll reinsert it in the _Lab_ list.

TurboGit reviewed (approved):
Ok, good to go to me. Thanks!

TurboGit commented:
Just waiting for the CI to end...

TurboGit commented:
Still failing but I don't think that it is related to this PR, at least the crash on final test is only on GCC 12 and I don't reproduce locally... Looks like a CI issue to me.

dterrahe commented:
@TurboGit the release note for this should refer to `processing modules/<focused>` as per the OP of this PR. Somehow the second bit of that got dropped in markdown, so it looks confusing now.

TurboGit commented:
> @TurboGit the release note for this should refer to `processing modules/<focused>` as per the OP of this PR. Somehow the second bit of that got dropped in markdown, so it looks confusing now.

Should be fixed, thanks for catching this!

wpferguson commented:
@dterrahe could you please check https://docs.darktable.org/lua/stable/lua.api.manual/darktable/gui/mimic/ and make it's correct.

Thanks
